1_fetch/src/get_wqp_inventory.R

```diff
@@ -0,0 +1,46 @@
+#' Inventory WQP within area of interest
+#' 
+#' @description Function to inventory WQP sites and observations within the area of interest
+#' 
+#' @param bbox_tbl data frame containing an attribute id and bounding box for each
+#' box sampled within the area of interest. The column bbox is of class list.
+#' @param char_names character string indicating the WQP CharacteristicNames to query
+#' @param site_type character string indicating which water feature type(s) to query
+#' @param year_earliest integer; the earliest year to include in the returned inventory,
+#' defaults to NULL, which returns all available years
+#' @param year_latest integer; the most recent year to include in the returned inventory,
+#' defaults to NULL, which returns all available years
+#' 
+#' @value returns a list of sites from the Water Quality Portal
+#' 
+#' @example inventory_wqp(aoi_poly, "Conductivity", "Stream")
+#' @example inventory_wqp(aoi_poly, "Temperature", "Lake, Reservoir, Impoundment")
+#' 
+
+inventory_wqp <- function(bbox_tbl, char_names, site_type, year_earliest = NULL, year_latest = NULL){
+  
+  # Return bounding box to use for WQP data summary
+  bbox <- bbox_tbl$bbox[[1]]
+  
+  # Inventory available WQP data
+  wqp_inventory <- dataRetrieval::readWQPsummary(
+    bBox = c(bbox$xmin, bbox$ymin, bbox$xmax, bbox$ymax),
+    summaryYears = "all",
+    siteType = site_type)
+  
+  # Format WQP inventory and subset years of interest
+  wqp_inventory_out <- wqp_inventory %>%
+    select(MonitoringLocationIdentifier, YearSummarized, OrganizationIdentifier, 
+           CharacteristicName, MonitoringLocationLongitude, 
+           MonitoringLocationLatitude, ResultCount) %>%
+    rename(lon = MonitoringLocationLongitude, 
+           lat = MonitoringLocationLatitude) %>%
+    filter(CharacteristicName %in% char_names) %>%
+    filter(if(!is.null(year_earliest)) YearSummarized >= year_earliest else TRUE) %>%
+    filter(if(!is.null(year_latest)) YearSummarized <= year_latest else TRUE) 
```

Comment:
@lekoenig do you understand what `YearSummarized` means? I don't think I get it

Reply:
Yeah, `YearSummarized` is an attribute field returned from `readWQPsummary`. The output data frame gives the `ResultCount` for every site identifier, characteristic name, and year. I added these filter steps based on `YearSummarized` so that we could subset the inventory based on a certain range of years. However, I think the simpler solution is to use another dataRetrieval function, probably `whatWQPdata()`. I'll circle back to this function once we've decided on our use/intention for the inventory target.

Reply:
In [7f4d4e0](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/21/commits/7f4d4e0953a5ebff6da1de72a7f1cc6bb74525b7), I've opted to call `whatWQPdata()` rather than the summary service since the arguments can match what a user would potentially use when querying the WQP.

Reply:
I don't know your specific inventory/long-term-goals, but you assuming you all are doing some crazy big pulls - you might want to do a few benchmarks to see which are more useful. 

`whatWQPdata` does give you more query options, that's true. However....there have been many cases historically where adding arguments to the query really increase the time it takes to get data back. Also, the information you get back from this function is only the number of samples/results with no breakdown of what kind of data it is or when it was taken. So, if you're interested in a single specific characteristic name - it's useful, if it's a bunch of characteristic names...it becomes a little harder to interpret. 

`readWQPsummary` gives you a better idea of how much data is available over time, and it's broken down by characteristic name (so if you have multiple names, you can still get a good idea of how much data there is for each). One other thing to think about with this function...if this was mostly what you wanted but you didn't want to post-filter out the site types or sample types, the WQP team might be agreeable to adding those features to the service. I think they can't do start/end dates because they don't want to deal with partial years...but site type or sample type might be worth bringing up to them....in fact....I think I'll do that anyway and let you know how easy/hard it would be.

Reply:
Thanks, @ldecicco-USGS, this added context is really useful! For this inventory, our goal is to summarize the data that is available on WQP and depending on how much data is available, use these summaries to decide whether to further discretize our WQP pulls into smaller chunks. To get around the limitation that `whatWQPdata` returns a single characteristic name, I've mapped this function over each of our desired characteristic names. That works for our use case in this example pipeline, but I'm also aware that this implementation could potentially really increase the time it takes to build the inventory target, especially if a user adds a long list of characteristic names or additional arguments.

And thank you for your offer to reach out to the WQP team! I wouldn't expect to want or need a lengthy list of options when calling the summary service. But I tend to think site type/sample type are common arguments used when pulling WQP data and in this case I wanted the inventory to accurately reflect the WQP data that match the intended search criteria.